feat: add onboard screens

toaderandrei opened the pull request:
* Add onboard screens to configure user profile,
ConfigureAddressFragment, ConfigureAccountFragment and their respective
view models.
* Add nav_graph for onboard screens.
* Add class to bind views from layout to objects.
* Add MapViewUtils to handle map related stuff.
* Add PermissionUtils to handle permissions related stuff.
* Add callbacks for click button events.
* Add google maps for localization.
* Add temporary icons used for the layouts.
* Add resources values as well.

Implements: #11, #13 

<!--
### Requirements for making a pull request

Thank you for contributing to our project!

Please fill out the template below to help the project maintainers review it as fast as possible and include your contribution to the project.
-->

### What does it fix?
<!--
If it is related to an open issue, please link the issue here. 
-->

Closes #11, #13 

<!-- Please mention the main changes this PR brings. -->

### How has it been tested?

<!-- Please describe the tests that you ran to verify your changes. -->

<!-- If applicable, mention any known issues with the pull request -->

Commit:
feat: add onboard screens

* Add onboard screens to configure user profile,
ConfigureAddressFragment, ConfigureAccountFragment and their respective
view models.
* Add nav_graph for onboard screens.
* Add class to bind views from layout to objects.
* Add MapViewUtils to handle map related stuff.
* Add PermissionUtils to handle permissions related stuff.
* Add callbacks for click button events.
* Add google maps for localization.
* Add temporary icons used for the layouts.
* Add resources values as well.

Implements: #11, #12

Commit:
feat: add onboard screens

* Add onboard screens to configure user profile,
ConfigureAddressFragment, ConfigureAccountFragment and their respective
view models.
* Add nav_graph for onboard screens.
* Add class to bind views from layout to objects.
* Add MapViewUtils to handle map related stuff.
* Add PermissionUtils to handle permissions related stuff.
* Add callbacks for click button events.
* Add google maps for localization.
* Add temporary icons used for the layouts.
* Add resources values as well.

Implements: #11, #12

aniri commented on app/src/main/java/ro/code4/deurgenta/bindings/BindingAdapters.kt:
```diff
@@ -0,0 +1,30 @@
+    @SuppressLint("ResourceType")
+    @JvmStatic
+    @BindingAdapter(value = ["imageRes"], requireAll = false)
+    fun bindImageResource(imageView: ImageView, resId: Int?) {
```
where is this used?

toaderandrei replied:
Correct, not used atm. It was used but then I decided to do it in the next PR. Will remove it.

aniri commented on app/src/main/AndroidManifest.xml:
```diff
@@ -4,6 +4,17 @@
+    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
+
+
+    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
+    <uses-permission android:name="android.permission.ACCESS_FINE_LOCATION" />
+    <uses-permission android:name="android.permission.BLUETOOTH" />
+    <uses-permission android:name="android.permission.BLUETOOTH_ADMIN"/>
+    <uses-permission android:name="android.permission.CHANGE_NETWORK_STATE" />
+    <uses-permission android:name="android.permission.CHANGE_WIFI_STATE" />
+    <uses-permission android:name="android.permission.READ_PHONE_STATE" android:maxSdkVersion="22" />
+    <uses-permission android:name="android.permission.WAKE_LOCK" />
```
please only add needed permissions

toaderandrei replied:
Ofc. Only needed permissions. This is the amount of permissions here sdk requires, at least in their sample apps. I will try to do some tests, maybe some can be removed. However, It might that some are specific to different android api levels.

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt:
```diff
@@ -0,0 +1,182 @@
+        val query = TextQuery(queryString, viewportGeoBox)
+
+        val maxItems = DEFAULT_MAX_ITEMS
+        val searchOptions = SearchOptions(LanguageCode.EN_GB, maxItems)
```
you could use DEFAULT_MAX_ITEMS directly here, no need for extra var

also, shouldn't the search language be set to ro?

toaderandrei replied:
Probably, but not sure it actually matters. I am searching in Austria, no difference seen. I will do some tests with country related code.

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt:
```diff
@@ -0,0 +1,182 @@
+        searchEngine.search(query, searchOptions, querySearchCallback)
+    }
+
+    fun searchOnMapAutoSuggestion(queryString: String) {
```
almost 100% duplicated code with method from above

aniri commented on app/src/main/java/ro/code4/deurgenta/ui/address/ConfigureAddressFragment.kt:
```diff
@@ -0,0 +1,198 @@
+        viewBinding.btSaveAddress.visibility = flag
+    }
+
+    private fun searchOnMap(queryString: String) {
+        mapViewUtils.searchOnMap(queryString)
+    }
+
+    private fun searchOnMapAutoSuggestion(queryString: String) {
+        mapViewUtils.searchOnMapAutoSuggestion(queryString)
+    }
+
+    fun dismissKeyboard(view: View) {
+        hideSoftInput(view)
+    }
```
what is the utility of these methods?

toaderandrei replied:
Basically it is about the hiding of the keyboard when you start the fragment for the first time.

aniri replied:
I meant why do we need these 3 methods since we could just use the methods from maputils/utils directly. the method from this class do not bring extra functionality

aniri commented on app/src/main/java/ro/code4/deurgenta/ui/base/Layout.kt:
```diff
@@ -0,0 +1,10 @@
+import androidx.annotation.LayoutRes
+import ro.code4.deurgenta.interfaces.ExcludeFromCodeCoverage
+
+interface Layout {
```
this interface already existed

aniri commented on app/src/main/java/ro/code4/deurgenta/ui/base/ViewModelSetter.kt:
```diff
@@ -0,0 +1,8 @@
+import androidx.lifecycle.ViewModel
+import ro.code4.deurgenta.interfaces.ExcludeFromCodeCoverage
+
+interface ViewModelSetter<out T : ViewModel> {
```
also duplicated

aniri commented on app/src/main/res/drawable/roundedbutton.xml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
```
could we reuse the buttons from auth activity?

aniri commented on app/src/main/res/layout/onboarding_configure_account.xml:
```diff
@@ -0,0 +1,99 @@
+                style="@style/MenuTextView"
+                android:background="?android:attr/selectableItemBackground"
+                android:onClick="@{() -> callbackAddresses.configure()}"
+                android:padding="12dp"
```
please use dimens

aniri commented on app/src/main/res/layout/onboarding_configure_addresses.xml:
```diff
@@ -0,0 +1,105 @@
+                android:layout_height="?attr/actionBarSize"
+                android:background="@android:color/transparent"
+                app:navigationIcon="@drawable/ic_close_24"
+                app:title="Add your own address"
```
extract to strings

aniri commented on app/src/main/res/layout/onboarding_configure_addresses.xml:
```diff
@@ -0,0 +1,105 @@
+                android:id="@+id/bt_locate_me"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:layout_marginTop="25dp"
```
dimens

aniri commented:
the app crashed when I started to write an address in the input field :( 
```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: ro.code4.deurgenta, PID: 7850
    java.lang.RuntimeException: GeoBox creation failed, corners are null.
        at ro.code4.deurgenta.helper.MapViewUtils.getMapViewGeoBox(MapViewUtils.kt:61)
        at ro.code4.deurgenta.helper.MapViewUtils.searchOnMapAutoSuggestion(MapViewUtils.kt:88)
        at ro.code4.deurgenta.ui.address.ConfigureAddressFragment.searchOnMapAutoSuggestion(ConfigureAddressFragment.kt:188)
        at ro.code4.deurgenta.ui.address.ConfigureAddressFragment.access$searchOnMapAutoSuggestion(ConfigureAddressFragment.kt:30)
        at ro.code4.deurgenta.ui.address.ConfigureAddressFragment$onQueryTextListener$1.onQueryTextChange(ConfigureAddressFragment.kt:160)
```

toaderandrei commented:
It could if the map was not initialized somehow. Will check this.

aniri commented on app/src/main/res/values/dimens.xml:
```diff
@@ -1,13 +1,31 @@
 <resources>
```
please refrain from adding too many different values and follow material guidelines https://material.io/archive/guidelines/layout/metrics-keylines.html#metrics-keylines-keylines-spacing
the layout doesn't have to be 100% aligned with the design

Commit:
fixup! feat: add onboard screens

remove hack to test the map

aniri commented on app/src/main/java/ro/code4/deurgenta/interfaces/ConfigureCallback.kt:
```diff
@@ -0,0 +1,5 @@
+package ro.code4.deurgenta.interfaces
+
+interface ConfigureCallback {
```
please rename this to something more clear, maybe configureaddresscallback?

aniri commented on app/src/main/res/layout/onboarding_configure_account.xml:
```diff
@@ -0,0 +1,110 @@
+                style="@style/MenuTextView"
+                android:background="?android:attr/selectableItemBackground"
+
+                android:padding="12dp"
```
please reuse values from dimens

aniri commented:
the app still crashes with the same error:
```
D/AndroidRuntime: Shutting down VM
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: ro.code4.deurgenta, PID: 4731
    java.lang.RuntimeException: GeoBox creation failed, corners are null.
        at ro.code4.deurgenta.helper.MapViewUtils.getMapViewGeoBox(MapViewUtils.kt:61)
        at ro.code4.deurgenta.helper.MapViewUtils.searchOnMap(MapViewUtils.kt:78)
        at ro.code4.deurgenta.helper.MapViewUtils.searchOnMap$default(MapViewUtils.kt:71)
        at ro.code4.deurgenta.ui.address.ConfigureAddressFragment$onQueryTextListener$1.onQueryTextChange(ConfigureAddressFragment.kt:165)
```

both on the phone and emulator.

could you please investigate?

Commit:
fixup! feat: add onboard screens

fix crash in android api versions below 30 by adding multiple permissions checks.

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt:
```diff
@@ -0,0 +1,275 @@
+                } else {
+                    Log.e(TAG, "error loading results:$errorCode")
+                    if (errorCode != MapError.OPERATION_IN_PROGRESS) {
+                        callback.onError(MapDataError("cannot load map. Error:$errorCode", MapErrorType.NOT_INITIALIZED))
```
please extract error strings in strings.xml

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt:
```diff
@@ -0,0 +1,275 @@
+            }
+        } else {
+            if (searchType == SearchType.STANDARD) {
+                callback.onError(MapDataError("Map is not initialized", MapErrorType.NOT_INITIALIZED));
```
please extract error strings in strings.xml

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt:
```diff
@@ -0,0 +1,275 @@
+
+                    }
+                    LocationSettingsStatusCodes.SETTINGS_CHANGE_UNAVAILABLE -> {
+                        val errorMessage = "Location settings are inadequate, and cannot be fixed here. Fix in Settings."
```
please extract error strings in strings.xml

aniri commented:
please extract all error texts that might be displayed to the users in the strings.xml files

aniri commented on app/build.gradle:
```diff
@@ -93,24 +97,15 @@ android {
-    firebaseMessagingVersion = "20.2.0"
-    firebaseConfigVersion = "19.1.4"
-    adapterDelegatesVersion = "4.2.0"
+    ndkVersion '20.0.5594570'
```
this seems to be added twice and with different versions

aniri commented on app/src/main/java/ro/code4/deurgenta/data/AppDatabase.kt:
```diff
@@ -8,16 +8,20 @@ import androidx.room.TypeConverters
-    entities = [Address::class, Backpack::class, BackpackItem::class],
-    version = 1
+    entities = [Backpack::class, BackpackItem::class, MapAddress::class],
+    version = 3
```
do we really need to increase the version?

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt:
```diff
@@ -0,0 +1,380 @@
+            longitude = searchResult.geoCoordinates!!.longitude,
+            fullAddress = searchResult.address.addressText
+        )
+        mapAddress.streetAddress = getStreetAddress(searchResult.address)
```
you could use formatAddress directly here, no need for another method..

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt:
```diff
@@ -0,0 +1,380 @@
+    }
+
+    private fun getMapViewGeoCircle(): GeoCircle {
+        return GeoCircle(geoCoordinates, 50000.0)
```
could you extract the radius value in the config files so it could be easily changed?

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/MapViewUtils.kt:
```diff
@@ -0,0 +1,380 @@
+        if (list.size == 1) {
+            loadMapSceneForGeoCoordinates(list[0].geoCoordinates)
+            callback.onSearchSuccess()
+        } else if (list.size > 1) {
```
I don't think there's a need for an explicit if condition here

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/PermissionManager.kt:
```diff
@@ -8,39 +8,34 @@ import androidx.fragment.app.Fragment
     }
 
-    fun requestPermissions(vararg permissionString: String) {
+    private fun checkPermissionsInternal(vararg permissionString: String) {
```
why was this renamed? the new name is not more clear...

aniri commented on app/src/main/java/ro/code4/deurgenta/helper/PermissionManager.kt:
```diff
@@ -49,18 +44,23 @@ open class PermissionManager(private val activity: Activity, private val fragmen
+
+    companion object {
+        @JvmStatic
+        val PERMISSION_REQUEST = 320
+
+        @JvmStatic
+        val PERMISSION_CHECK_SETTINGS = 320
```
shouldn't these have different values? 🤔

aniri commented on app/src/main/java/ro/code4/deurgenta/interfaces/ViewCallback.kt:
```diff
@@ -0,0 +1,9 @@
+/**
+ * Callback for when saving progress.
+ */
+interface ViewCallback {
```
could you please rename it to something more clear?

aniri commented on app/src/main/res/layout/fragment_configure_account.xml:
```diff
@@ -0,0 +1,117 @@
+                    style="@style/MenuTextView"
+                    android:background="?android:attr/selectableItemBackground"
+                    android:onClick="@{() -> callbackAddresses.call()}"
+                    android:padding="12dp"
```
please use values form dimens file

aniri commented on app/src/main/res/layout/fragment_configure_account.xml:
```diff
@@ -0,0 +1,117 @@
+                    style="@style/MenuTextView"
+                    android:background="?android:attr/selectableItemBackground"
+
+                    android:padding="12dp"
```
dimens

aniri commented on app/src/main/res/values/styles.xml:
```diff
@@ -61,13 +61,26 @@
         <item name="lineHeight">24sp</item>
     </style>
 
+    <style name="Text.OptionItem">
```
i think you could have reused Text.MenuList

aniri commented on app/src/main/res/values/styles.xml:
```diff
@@ -61,13 +61,26 @@
         <item name="lineHeight">26sp</item>
     </style>
 
+    <style name="Text.Content">
```
i think you could have reused Text.Default

Commit:
fixup! feat: add onboard screens

process code review comments and fix bug

aniri committed:
extract search radius in config files

aniri committed:
cleaned up code
